### PR TITLE
Bump to Cilium 1.14

### PR DESCRIPTION
**Pull request by zadjadr**
I would love some help with the PR.

Stuff still missing: 

- [x] Manifest hashes (how to get them?)
- [x] First upgrade Cilium to v1.13
- [x] Deploy working cluster using roling update
- [x] Deploy working cluster from scratch
- [x] Allow wireguard + L7 proxy
- [x] Test hubble-relay

Can't test myself:

- [ ] AWS ENI stuff (new lifecycle.postStart for cilium-agent)

Better done in separate PRs: 
- [ ] Test wireguard + L7 proxy
- [ ] Add new functionality (e.g L2 Announcement, bgp, etc.)

Not 100% related changes:

- Switched from mounting /var/run/cilium in the hubble-relay deployment to using a service and exposing the hubble-peer port in the cilium agent (better for security, better practice etc.)

### CODE REVIEW

**Comment by hakman** [comment]
> Manifest hashes (how to get them?)

Run hack/update-expected.sh.
Also please squash the commits.

**Comment by zadjadr** [comment]
I'm getting this error on the cilium-operator:

```
cilium-operator level=fatal msg="Option --routing-mode cannot be used in combination with --tunnel" subsys=config
```

Maybe someone knows what I'm missing.

I tried both configs

- `spec.networking.cilium.tunnel=disabled` (which should replace `tunnel` with `routing-mode: "native"` in the configmap
- `spec.networking.cilium.tunnel=vxlan`

**Comment by olemarkus** [comment]
This could be an upstream bug. I see cilium-operator complaining about what you mention above, but looking at https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/15705/pull-kops-e2e-cni-cilium-etcd/1684943371451240448/artifacts/cluster-info/kube-system/configmaps.yaml I don't see routing mode set.

**Comment by hakman** [comment]
@zadjadr Let's try with Cilium 1.13 first. WDYT?

**Comment by zadjadr** [comment]
@hakman I'll give it a try!

**Comment by zadjadr** [comment]
@hakman I tried with 1.13.5, here is what I encountered:

- Updating worker nodes works
- Updating control plane nodes does not work

The operator and agent version can be updated to 1.13.5, new worker nodes will join the cluster without issues.

A new control plane node is not able to connect to the Cluster:

```
==> /var/log/kube-apiserver.log <==
}. Err: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4001: connect: connection refused"
```

`etcd` (and some other services) are never started. No idea why this happens, since it should not be related to cilium.. I'll give it another try later when I have the time.

**Comment by hakman** [comment]
@zadjadr It is unrelated, probably related to the fact that you are testing with master. Try building based on the release-1.27 branch

**Comment by zadjadr** [comment]
@hakman worked! Thanks, I'll create a [new PR](https://github.com/kubernetes/kops/pull/15730) to first bump to 1.13 and whenever 1.14 is ready, we can merge this PR.

**Comment by zadjadr** [comment]
v1.14.0 is now almost working fine. The agent and the operator are deployed and running. The issue I encounter now is that the node never becomes ready, because the "cni is not ready" (reported by kubelet on the node). 

I spoke to the cilium folks and they mentioned it would be better to wait for v1.14.1 or v1.14.2 before integrating it into Kops, so I'll wait until then try some more debugging until then.

**Comment by hakman** [comment]
> I spoke to the cilium folks and they mentioned it would be better to wait for v1.14.1 or v1.14.2 before integrating it into Kops, so I'll wait until then try some more debugging until then.

We are planning to release kOps 1.28 in Sept, so this would probably not make it there.
Should be good timing to include it in kOps 1.29.

**Comment by zadjadr** [comment]
@hakman @olemarkus I am now able to deploy and roll a cluster with Cilium v1.14.0 :tada: 

Some refining might be needed, so I created two commits for now - I'll squash in the end if you like.
Some of the new features are not yet implemented, but I don't know if we want to implement them (e.g. L2 Announcements) in this PR or via new PRs as needed.

swappy-20230808_124447

swappy-20230808_124503

swappy-20230808_124947

**Comment by hakman** [comment]
Please don't remove `seLinuxOptions`.

**Comment by zadjadr** [comment]
Everything (including hubble relay) is working now. I will try out wireguard + L7 proxy etc. now and finalize the PR.

```
$ cilium status                              
    /¯¯\
 /¯¯\__/¯¯\    Cilium:             OK
 \__/¯¯\__/    Operator:           OK
 /¯¯\__/¯¯\    Envoy DaemonSet:    disabled (using embedded mode)
 \__/¯¯\__/    Hubble Relay:       OK
    \__/       ClusterMesh:        disabled

DaemonSet              cilium             Desired: 6, Ready: 6/6, Available: 6/6
Deployment             hubble-ui          Desired: 1, Ready: 1/1, Available: 1/1
Deployment             cilium-operator    Desired: 2, Ready: 2/2, Available: 2/2
Deployment             hubble-relay       Desired: 2, Ready: 2/2, Available: 2/2
Containers:            cilium             Running: 6
                       hubble-ui          Running: 1
                       cilium-operator    Running: 2
                       hubble-relay       Running: 2
Cluster Pods:          10/10 managed by Cilium
Helm chart version:    1.14.0
Image versions         hubble-relay       quay.io/cilium/hubble-relay:v1.14.0@sha256:bfe6ef86a1c0f1c3e8b105735aa31db64bcea97dd4732db6d0448c55a3c8e70c: 2
                       cilium             quay.io/cilium/cilium:v1.14.0@sha256:5a94b561f4651fcfd85970a50bc78b201cfbd6e2ab1a03848eab25a82832653a: 12
                       hubble-ui          quay.io/cilium/hubble-ui:v0.12.0@sha256:1c876cfa1d5e35bc91e1025c9314f922041592a88b03313c22c1f97a5d2ba88f: 1
                       hubble-ui          quay.io/cilium/hubble-ui-backend:v0.12.0@sha256:8a79a1aad4fc9c2aa2b3e4379af0af872a89fcec9d99e117188190671c66fc2e: 1
                       cilium-operator    quay.io/cilium/operator:v1.14.0@sha256:0fad2cce5d17221e90538dfff552c402aae93266ab8ce531b04f722f407f71a8: 2
```

The only issue I currently see: cilium-agent is running as privileged although it should be running with a specific security context..
We were using privileged before as well, but I'd like to use a least privilege approach if possible.
I'll give it some more thought & in the worst case we can improve that after merging this PR, if you are fine with it.

**Comment by hakman** [comment]
@zadjadr We started working on kOps 1.29, so this can be merged once the review is finished.
Please rebase, update expected files and fix the tests to continue the review.

**Comment by zadjadr** [comment]
@hakman can you help me understand why the two tests are failing? At least for one of them I see `system-cluster-critical pod "cilium-operator-c65f88f75-btj9c" is not ready (cilium-operator)` but are there any logs I could check out that provide more info?

I am able to deploy a cluster in Openstack without any issues.

**Comment by zadjadr** [comment]
> @hakman can you help me understand why the two tests are failing? At least for one of them I see `system-cluster-critical pod "cilium-operator-c65f88f75-btj9c" is not ready (cilium-operator)` but are there any logs I could check out that provide more info?
> 
> I am able to deploy a cluster in Openstack without any issues.

Okay found some logs [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/15705/pull-kops-e2e-cni-cilium-eni/1698754581581795328/artifacts/cluster-info/kube-system/cilium-operator-7f5989f9d5-spjfn/logs.txt)

The only concerning line is:

```
level=warning msg="Failed to update CiliumNode" attempt=0 error="Operation cannot be fulfilled on ciliumnodes.cilium.io \"i-0c501eb0b73a3f997\": the object has been modified; please apply your changes to the latest version and try again" instanceID=i-0c501eb0b73a3f997 name=i-0c501eb0b73a3f997 subsys=ipam updateStatus=true
```

And the only real change to the cilium-operator is [adding the readiness probe](https://github.com/kubernetes/kops/pull/15705/files#diff-c52a919012cae20096b59b47fb10994a481f3b47d0c1513c3a7d7e68973b4656R1357-R1366) & doing the probes via ipv6 if IPv6Only is activated.

**Comment by zadjadr** [comment]
Looks like removing the `readinessprobe` fixed one of the tests. I don't know if `pull-kops-e2e-cni-cilium-eni` is just flaky or if my changes make if fail - do you have any idea how to debug this? @hakman

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template:768]: I think `OnDelete` was done on purpose.

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template:1157]: We want this set as node affinity.
```suggestion
      nodeSelector: null
```

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template:671]: I think this was on purpose and we want it back

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template:1489]: Same as other node selector

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by hakman** [comment]
> Looks like removing the `readinessprobe` fixed the failing `cilium-operator` pods. I don't know if `pull-kops-e2e-cni-cilium-eni` is just flaky or if my changes make if fail - do you have any idea how to debug this? @hakman

Doesn't seem to be flaky, so may be the changes or Cilium bug.
To debug, open the Details link on the failed test and navigate to:
* Artifacts (top right corner) / artifacts / cluster-info / kube-system / cilium for pod logs
* Artifacts (top right corner) / artifacts / i-* for node logs.

**Comment by zadjadr** [comment]
> > Looks like removing the `readinessprobe` fixed the failing `cilium-operator` pods. I don't know if `pull-kops-e2e-cni-cilium-eni` is just flaky or if my changes make if fail - do you have any idea how to debug this? @hakman
> 
> Doesn't seem to be flaky, so may be the changes or Cilium bug. To debug, open the Details link on the failed test and navigate to:
> 
> * Artifacts (top right corner) / artifacts / cluster-info / kube-system / cilium for pod logs
> * Artifacts (top right corner) / artifacts / i-* for node logs.

Hmm.. all the cilium and cilium-operator I checked, look fine. The kubelet seems to have issues, but I don't know where the issue might be.

I see the cilium pods are marked as ready as well:

```log
I0910 22:45:21.829982    3139 kubelet.go:2493] "SyncLoop (probe)" probe="startup" status="started" pod="kube-system/cilium-thp78"
I0910 22:45:21.832634    3139 kubelet.go:2493] "SyncLoop (probe)" probe="readiness" status="ready" pod="kube-system/cilium-thp78"
I0910 22:45:21.851564    3139 pod_startup_latency_tracker.go:102] "Observed pod startup duration" pod="kube-system/cilium-thp78" 
```

But looks like the [`aws-node-termination-handler`](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/15705/pull-kops-e2e-cni-cilium-eni/1700999486857285632/artifacts/cluster-info/kube-system/aws-node-termination-handler-5d479c96f5-l4hfd/logs.txt) does not find the nodes.

The `cilium-operator` should have set removed the not ready labels, so I don't know where to check further.

```
level=info msg="Managing Cilium Node Taints or Setting Cilium Is Up Condition for Kubernetes Nodes" k8sNamespace=kube-system label-selector="k8s-app=cilium" remove-cilium-node-taints=true set-cilium-is-up-condition=true set-cilium-node-taints=true subsys=cilium-operator
level=info msg="Starting to synchronize CiliumNode custom resources" subsys=cilium-operator
level=info msg="Starting to garbage collect stale CiliumNode custom resources" subsys=watchers
level=info msg="Starting to garbage collect stale CiliumEndpoint custom resources" subsys=cilium-operator
level=info msg="Cilium pod scheduled but not running for node; setting taint" nodeName=i-0fabbea984aec141e subsys=watchers
level=info msg="Garbage collected status/nodes in Cilium Network Policies found=0, gc=0" subsys=cilium-operator
level=info msg="Garbage collected status/nodes in Cilium Clusterwide Network Policies found=0, gc=0" subsys=cilium-operator
level=info msg="CiliumNodes caches synced with Kubernetes" subsys=cilium-operator
level=info msg="Discovered new CiliumNode custom resource" name=i-0fabbea984aec141e subsys=ipam
level=info msg="Starting CNP derivative handler" subsys=cilium-operator
level=info msg="Starting CCNP derivative handler" subsys=cilium-operator
level=info msg="Initialization complete" subsys=cilium-operator
```

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.14.yaml.template:214]: I think this is a duplicated at line 322.
```suggestion
```

**Comment by zadjadr** [comment]
No idea why `pull-kops-e2e-cni-cilium-eni`. Might be some issue in my changes or something in upstream cilium. If someone has an idea, I'd be happy to read/change stuff in this PR.

**Comment by olemarkus** [comment]
Hard to say. Tried to bring up this branch and it fails with at least service routing to the API server. I would ask the cilium channel why this may happen.

**Comment by zadjadr** [comment]
@olemarkus thanks, I asked in the [cilium slack channel](https://cilium.slack.com/archives/C53TG4J4R/p1695470175398899) - I'll rebase the PR if we get some more hints.

Personally I'm running this branch in Openstack without any issues & the tests only fail for eni so I assume the issue must be somewhere around here: https://github.com/kubernetes/kops/pull/15705/files#diff-c52a919012cae20096b59b47fb10994a481f3b47d0c1513c3a7d7e68973b4656R1039-R1065

**Comment by hakman** [comment]
We discussed yesterday and agreed to do some more checks next week and make the eni test optional if we cannot find the issue.

**Comment by zadjadr** [comment]
@hakman Any updates on this?

**Review by hakman — APPROVED** [review]

**Comment by hakman** [comment]
Thanks @zadjadr! 🥳

**Comment by rifelpet** [comment]
There are E2E failures for cilium-eni that started after this was merged.

Examples:
* https://testgrid.k8s.io/kops-grid#kops-grid-cilium-eni-u2204-k28
* https://testgrid.k8s.io/kops-grid#kops-grid-cilium-eni-u2004-k28

Any chance someone can take a look?

**Pull request merged**
